### PR TITLE
366:  validate asset keys

### PR DESCRIPTION
**Pull request by ahalberkamp**
In This pr, it is validated that assetKeys exists on old and new resources exist.If not a warning-callback is triggered
Fixes: https://github.com/commercetools/commercetools-sync-java/issues/366

### CODE REVIEW

File: src/main/java/com/commercetools/sync/commons/utils/AssetsUpdateActionUtils.java [hunk 1]
```diff
@@ -21,14 +25,21 @@
...
 import static java.util.Optional.ofNullable;
 import static java.util.stream.Collectors.toCollection;
 import static java.util.stream.Collectors.toList;
 import static java.util.stream.Collectors.toMap;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+
 
 public final class AssetsUpdateActionUtils {
 
+    private static final String ASSET_KEY_NOT_SET = "Asset with %s has no key set. Keys are required for "
```
Comment (JudeNiroshan) [inline src/main/java/com/commercetools/sync/commons/utils/AssetsUpdateActionUtils.java:39]:
```suggestion
    private static final String ASSET_KEY_NOT_SET = "Asset with %s has no defined key. Keys are required for "
```

File: src/test/java/com/commercetools/sync/categories/utils/categoryupdateactionutils/BuildAssetsUpdateActionsTest.java [hunk 2]
```diff
@@ -89,21 +101,89 @@ void buildAssetsUpdateActions_WithNewAssetsAndNoOldAssets_ShouldBuild3AddActions
...
+                    .key("a").tags(emptySet()).build(), 0),
             AddAsset.of(
                 AssetDraftBuilder.of(singletonList(AssetSourceBuilder.ofUri("uri").build()), ofEnglish("asset name"))
-                                 .key("c").tags(emptySet()).build(), 2)
+                    .key("b").tags(emptySet()).build(), 1)
+
+
         );
+        assertThat(warningCallBackMessages.get(0)).isEqualTo("Asset with Name: LocalizedString(en -> asset name) "
+            + "has no key set. Keys are required for asset matching.");
```
Comment (JudeNiroshan) [inline src/test/java/com/commercetools/sync/categories/utils/categoryupdateactionutils/BuildAssetsUpdateActionsTest.java:141]:
```suggestion
            + "has no defined key. Keys are required for asset matching.");
```

File: src/test/java/com/commercetools/sync/categories/utils/categoryupdateactionutils/BuildAssetsUpdateActionsTest.java [hunk 3]
```diff
@@ -89,21 +101,89 @@ void buildAssetsUpdateActions_WithNewAssetsAndNoOldAssets_ShouldBuild3AddActions
...
+        final Category oldCategory = readObjectFromResource(CATEGORY_DRAFT_WITH_ASSETS_ABC, Category.class);
+        final CategoryDraft newCategoryDraft = readObjectFromResource(CATEGORY_DRAFT_WITH_ASSETS_WITHOUT_KEY,
+            CategoryDraft.class);
+
+        final List<UpdateAction<Category>> updateActions =
+            buildAssetsUpdateActions(oldCategory, newCategoryDraft, syncOptions);
+
+        assertThat(updateActions).containsExactlyInAnyOrder(
+            RemoveAsset.ofKey("c")
+
```
Comment (JudeNiroshan) [inline src/test/java/com/commercetools/sync/categories/utils/categoryupdateactionutils/BuildAssetsUpdateActionsTest.java:157]: unnecessary new line

File: src/test/java/com/commercetools/sync/categories/utils/categoryupdateactionutils/BuildAssetsUpdateActionsTest.java [hunk 4]
```diff
@@ -89,21 +101,89 @@ void buildAssetsUpdateActions_WithNewAssetsAndNoOldAssets_ShouldBuild3AddActions
...
+        );
+        assertThat(warningCallBackMessages.get(0)).isEqualTo("Asset with Name: LocalizedString(en -> asset name) "
+            + "has no key set. Keys are required for asset matching.");
```
Comment (JudeNiroshan) [inline src/test/java/com/commercetools/sync/categories/utils/categoryupdateactionutils/BuildAssetsUpdateActionsTest.java:160]:
```suggestion
            + "has no defined key. Keys are required for asset matching.");
```

File: src/main/java/com/commercetools/sync/categories/utils/CategoryUpdateActionUtils.java [hunk 5]
```diff
@@ -251,7 +251,8 @@
             return AssetsUpdateActionUtils.buildAssetsUpdateActions(
                 oldCategory.getAssets(),
                 newCategory.getAssets(),
-                new CategoryAssetActionFactory(syncOptions));
+                new CategoryAssetActionFactory(syncOptions),
```
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/categories/utils/CategoryUpdateActionUtils.java:254]: could be formatted better for readability.

File: src/main/java/com/commercetools/sync/commons/helpers/AssetActionFactory.java [hunk 6]
```diff
@@ -28,7 +28,7 @@
                                                             @Nonnull AssetDraft newAssetDraft);
 
     /**
-     * Takes an asset key to build a RemoveAsset action of the type T.
+     * Takes an asset  to build a RemoveAsset action of the type T.
```
[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/commons/helpers/AssetActionFactory.java:31]: I assume it was accidentally deleted 🤔

File: src/main/java/com/commercetools/sync/commons/utils/AssetsUpdateActionUtils.java (hunk 1 above)
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/commons/utils/AssetsUpdateActionUtils.java]: I guess, it's better to use the same convention/similar messages with the other errors, so I would go with Andi's message.
Another example for Category key, `Category with id: '%s' has no key set. Keys are required for category matching`

File: src/main/java/com/commercetools/sync/commons/utils/AssetsUpdateActionUtils.java [hunk 7]
```diff
@@ -81,25 +98,42 @@
...
-        final Map<String, Asset> oldAssetsKeyMap = oldAssets.stream().collect(toMap(Asset::getKey, asset -> asset));
+        final Map<String, Asset> oldAssetsKeyMap = new HashMap<>();
 
-        final Map<String, AssetDraft> newAssetDraftsKeyMap;
+        oldAssets.forEach(asset -> {
+            String assetKey = asset.getKey();
+            if (isNotBlank(assetKey)) {
+                oldAssetsKeyMap.put(assetKey, asset);
+            } else {
+                syncOptions.applyWarningCallback(format(ASSET_KEY_NOT_SET, "Id: " + asset.getId()));
```
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/commons/utils/AssetsUpdateActionUtils.java:117]:
```suggestion
                syncOptions.applyWarningCallback(format(ASSET_KEY_NOT_SET, "id: " + asset.getId()));
```

File: src/main/java/com/commercetools/sync/commons/utils/AssetsUpdateActionUtils.java [hunk 8]
```diff
@@ -81,25 +98,42 @@
...
+                if (isNotBlank(assetKey)) {
+                    newAssetDraftsKeyMap.merge(assetKey, newAsset, (assetDraftA, assetDraftB) -> {
                             throw new DuplicateKeyException("Supplied asset drafts have duplicate keys. Asset keys are"
                                 + " expected to be unique inside their container (a product variant or a category).");
                         }
-                    ));
+                    );
+
+                } else {
+                    syncOptions.applyWarningCallback(format(ASSET_KEY_NOT_SET, "Name: " + newAsset.getName()));
```
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/commons/utils/AssetsUpdateActionUtils.java:133]:
```suggestion
                    syncOptions.applyWarningCallback(format(ASSET_KEY_NOT_SET, "name: " + newAsset.getName()));
```

File: src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java [hunk 9]
```diff
@@ -253,7 +253,8 @@
             return buildAssetsUpdateActions(
                 oldProductVariant.getAssets(),
                 newProductVariant.getAssets(),
-                new ProductAssetActionFactory(oldProductVariant.getId(), syncOptions));
+                new ProductAssetActionFactory(oldProductVariant.getId(), syncOptions),
```
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java:256]: could be formatted better.

File: src/test/java/com/commercetools/sync/categories/utils/categoryupdateactionutils/BuildAssetsUpdateActionsTest.java [hunk 10]
```diff
@@ -192,14 +272,14 @@ void buildAssetsUpdateActions_WithOneExtraAsset_ShouldBuildAddAssetAction() {
...
-            buildAssetsUpdateActions(oldCategory, newCategoryDraft, SYNC_OPTIONS);
+            buildAssetsUpdateActions(oldCategory, newCategoryDraft, syncOptions);
 
         assertThat(updateActions).containsExactly(
             AddAsset.of(
                 AssetDraftBuilder.of(singletonList(AssetSourceBuilder.ofUri("uri").build()), ofEnglish("asset name"))
-                                 .key("d")
-                                 .tags(emptySet())
-                                 .build(), 3));
+                    .key("d")
```
Comment (ahmetoz) [inline src/test/java/com/commercetools/sync/categories/utils/categoryupdateactionutils/BuildAssetsUpdateActionsTest.java:280]: please try not to change code just for formatting, this is overriding the git history.

File: src/test/java/com/commercetools/sync/products/utils/ProductSyncUtilsTest.java [hunk 11]
```diff
@@ -96,8 +96,8 @@ void buildActions_FromDraftsWithMultipleDifferentValues_ShouldBuildUpdateActions
         final ProductDraftBuilder draftBuilder = createProductDraftBuilder(
             PRODUCT_KEY_1_CHANGED_WITH_PRICES_RESOURCE_PATH, ProductType.referenceOfId("anyProductType"));
 
-        final AssetDraft assetDraft = AssetDraftBuilder.of(singletonList(AssetSourceBuilder.ofUri("uri").build()),
-            ofEnglish("assetName")).build();
+        AssetDraft assetDraft = AssetDraftBuilder.of(singletonList(AssetSourceBuilder.ofUri("uri").build()),
```
Comment (ahmetoz) [inline src/test/java/com/commercetools/sync/products/utils/ProductSyncUtilsTest.java:99]: removing final why?

**Comment by ahalberkamp** [comment]
I have allready write some integration test, see: https://github.com/commercetools/commercetools-sync-java/pull/509/files#diff-cadc851f237ef9842d0224c27748bf3aR120 and I fixed the build @ahmetoz

File: src/main/java/com/commercetools/sync/commons/helpers/AssetActionFactory.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ahalberkamp) [reply src/main/java/com/commercetools/sync/commons/helpers/AssetActionFactory.java]: 91580ca5

File: src/main/java/com/commercetools/sync/commons/utils/AssetsUpdateActionUtils.java [hunk 12]
```diff
@@ -81,25 +98,42 @@
...
+        oldAssets.forEach(asset -> {
+            String assetKey = asset.getKey();
+            if (isNotBlank(assetKey)) {
+                oldAssetsKeyMap.put(assetKey, asset);
+            } else {
+                syncOptions.applyWarningCallback(format(ASSET_KEY_NOT_SET, "Id: " + asset.getId()));
+            }
+        });
+        final Map<String, AssetDraft> newAssetDraftsKeyMap = new HashMap<>();
+        ;
```
[the remark this answers is not in the document]
Comment (ahalberkamp) [reply src/main/java/com/commercetools/sync/commons/utils/AssetsUpdateActionUtils.java]: 738c1864

File: src/main/java/com/commercetools/sync/commons/utils/AssetsUpdateActionUtils.java [hunk 13]
```diff
@@ -1,16 +1,20 @@
 package com.commercetools.sync.commons.utils;
 
+
```
[the remark this answers is not in the document]
Comment (ahalberkamp) [reply src/main/java/com/commercetools/sync/commons/utils/AssetsUpdateActionUtils.java]: abd3ae1f

File: src/test/java/com/commercetools/sync/categories/utils/categoryupdateactionutils/BuildAssetsUpdateActionsTest.java (hunk 3 above)
Comment (ahalberkamp) [reply src/test/java/com/commercetools/sync/categories/utils/categoryupdateactionutils/BuildAssetsUpdateActionsTest.java]: adb7d0c0

File: src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java (hunk 9 above)
Comment (ahalberkamp) [reply src/main/java/com/commercetools/sync/products/utils/ProductVariantUpdateActionUtils.java]: 954b98ce

File: src/main/java/com/commercetools/sync/categories/utils/CategoryUpdateActionUtils.java (hunk 5 above)
Comment (ahalberkamp) [reply src/main/java/com/commercetools/sync/categories/utils/CategoryUpdateActionUtils.java]: 954b98ce

File: src/test/java/com/commercetools/sync/categories/utils/categoryupdateactionutils/BuildAssetsUpdateActionsTest.java (hunk 10 above)
Comment (ahalberkamp) [reply src/test/java/com/commercetools/sync/categories/utils/categoryupdateactionutils/BuildAssetsUpdateActionsTest.java:277]: I only fixed it according to checkstyle

File: src/test/java/com/commercetools/sync/products/utils/ProductSyncUtilsTest.java (hunk 11 above)
Comment (ahalberkamp) [reply src/test/java/com/commercetools/sync/products/utils/ProductSyncUtilsTest.java]: 117b4623

**Comment by butenkor** [comment]
@ahalberkamp Does it call warning callback and skip processing faulty resource or it will make a request to the API with assets without keys?

**Comment by ahalberkamp** [comment]
@butenkor assets without keys will be skipped, so no invalid request will be sent to the API.

**Comment by butenkor** [comment]
Ok, Thanks. What about sync statistics? Will the totals (updated, processed, failed, skipped...) be updated accordingly?

**Comment by ahalberkamp** [comment]
No, because if a assetkey is missing, the given Product will still be updated and only a Warning, that the asset is not synced, will be printed out. So there is no need to change the statistic in this case.

**Comment by butenkor** [comment]
ok. Thanks

**Pull request merged**